airflow_dags/wikipedia/ngrams/dolt_load.py

```diff
@@ -0,0 +1,354 @@
+    if article_count % TRI_SHARD_LEN == 0:
+        csv_num = int(article_count / TRI_SHARD_LEN)
+        write_to_csv(str(csv_num), 'trigram')
+    art_dicts = count_article_ngrams(article)
+    add_article_ngram_counts(art_dicts)
+
+
+def get_ngram_df_builder(ngram_name: str, lower: str) -> Callable[[], pd.DataFrame]:
+
+    def inner() -> pd.DataFrame:
+        df = pd.read_csv('all_{}s{}.csv'.format(ngram_name, lower))
+        return df.astype({'total_count': 'int', 'article_count': 'int'})
+
+    return inner
+
+
+def get_counts_df_builder(date_string: str, article_total: int):
+    uni_df = pd.read_csv('all_unigrams.csv')
+    word_total = uni_df['total_count'].sum()
+
+    def inner() -> pd.DataFrame:
+        df = pd.DataFrame([{'dump_date': date_string,
+                            'total_word_count': word_total,
+                            'total_article_count': article_total}])
+        return df.astype({'total_word_count': 'int', 'total_article_count': 'int'})
+
+    return inner
+
+
+def get_writers(date_string: str, article_count: int, lower=''):
+    writers = []
+
+    for ngram_name in NGRAM_DICTS.keys():
+        logging.info('Starting merge for {}s'.format(ngram_name))
+        merge_csvs(ngram_name, lower)
+        logging.info('Successfully merged all {} csvs'.format(ngram_name))
+        table_name = ngram_name + '_counts'
+        writers.append(get_df_table_writer(table_name,
+                                           get_ngram_df_builder(ngram_name, lower),
+                                           [ngram_name],
+                                           import_mode='replace'))
+
+    writers.append(get_df_table_writer('total_counts',
+                                       get_counts_df_builder(date_string, article_count),
+                                       ['dump_date'],
+                                       import_mode='update'))
+
+    return writers
+
+
+def get_dolt_datasets(date_string: str, dump_target: str):
+    """
+    Gets case-sensitive and case-insensitive loader for each Wikipedia dump
+    Each loader has a writer for each table
+    :param date_string:
+    :param dump_target:
+    :return:
+    """
+    loaders = []
+    article_count = fetch_data(dump_target)
+
+    # Get case-sensitive ngrams
+    writers = get_writers(date_string, article_count)
+    message = 'Update case-sensitive ngrams for dump date {}'.format(date_string)
+    loaders.append(get_dolt_loader(writers, True, message))
+    loaders.append(get_branch_creator('{}/case-sensitive'.format(date_string)))
+
+    # Get case-insensitive ngrams
+    shard_len = int(3e7)
+    for ngram_name in NGRAM_DICTS.keys():
+        get_lowered_ngrams(ngram_name, shard_len)
+    l_message = 'Update case-insensitive ngrams for dump date {}'.format(date_string)
+    l_writers = get_writers(date_string, article_count, lower='_lower')
+    loaders.append(get_dolt_loader(l_writers, True, l_message, '{}/case-insensitive'.format(date_string)))
```

Comment:
This is a much nicer API than we had before!

Reply:
definitely!